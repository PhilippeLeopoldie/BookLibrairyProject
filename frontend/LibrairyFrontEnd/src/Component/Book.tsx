import React, {useEffect, useState} from 'react'
 import {BookType} from '../Type'
 import FetchApi from '../FetchApi'
import PostOpinion from './PostOpinion'
 

 
function Books () {
    const [books, setBooks] = useState<BookType[]>()
    useEffect(() =>{
        FetchApi('http://localhost:5133/api/Books').then(books => setBooks(books))
    },[])
 //console.log(books?.at(Math.floor(Math.random()*books.length)))
 let random = books?.at(Math.floor(Math.random()*books.length))

    return (
        <>
        <h2>Suggestion of the day</h2>
        <h3>{random?.title}</h3>
        {
             books?.map((book, index)=>(
                <p key={index}>{book.title}{book.author}
                <PostOpinion bookId = {book.bookId}/></p>
                   
            )) 
            
        }
    </>
    )
}
 export default Books
