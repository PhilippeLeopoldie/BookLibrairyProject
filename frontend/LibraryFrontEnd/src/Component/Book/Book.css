.booksContainer {
  gap: 2%;
  display: grid;
  justify-content: center;
}

.bookcard--grid {
  border: 1px gray solid;
  border-radius: 3%;
  display: grid;
  grid-template-rows: 2fr 3fr 1fr;
  width: 20rem;
  height: 30rem;
  border-radius: 20px 16px 12px 32px;
  background-image: linear-gradient(
    to right,
    #d4af37 10%,
    rgb(253, 229, 187) 10%
  );
}
.bookCard__RateButton{
 margin-top :4%;
 margin-bottom: 8%;
 margin-left: 30%;
 margin-right: 20%;
 font-size:large;
 font-weight:bolder;
  color:saddlebrown; 
 background-color: transparent;
 border-color: #a38729;

}

.bookcard--grid:after {
  display: flex;
  justify-content: end;
  margin-left: 2%;
  height: 52px;
  bottom: 6px;
  right: 0px;
  background: white;
  border-radius: 32px 4px 4px 32px;
  box-shadow: inset 4px 6px 0px 0px #a69650;
  background-image: linear-gradient(
    to bottom,
    transparent 6px,
    #a69650 8px,
    transparent 8px,
    transparent 12px,
    #a69650 12px,
    transparent 14px,
    transparent 18px,
    #a69650 18px,
    transparent 20px,
    transparent 24px,
    #a69650 24px,
    transparent 26px,
    transparent 30px,
    #a69650 30px,
    transparent 32px,
    transparent 36px,
    #a69650 36px,
    transparent 38px,
    transparent 42px,
    #a69650 42px,
    transparent 44px,
    transparent 48px,
    #a69650 48px,
    transparent 50px
  );
  content: "";
  position: relative;
}

.bookcard--header {
  border-top-right-radius: 5%;
  border-top-left-radius: 5%;
  background-color: transparent;
  padding-left: 10%;
}

.booktitle {
  background-color: transparent;
  font-size: x-large;
  font-weight: bold;
  -webkit-text-stroke: 1px #a38729;
  color: #d4af37;
  margin: 0;
}

.bookauthor {
  background-color: transparent;
  margin:2%;
  margin-bottom: 8%;
}

.bookform--view{
  resize: none;
}

.booktitle--trashbutton {
  display: flex;
  width: 40px;
  height: 40px;
  margin: 2%;
}

.bookcard--iconeTrash {
  border-radius: 20%;
}

@media screen and (min-width: 395px) {
  .bookcard {
    width: 20rem;
  }
}

@media screen and (min-width: 698px) {
  .booksContainer {
    grid-template-columns: 1fr 1fr;
  }
}

@media screen and (min-width: 1042px) {
  .booksContainer {
    grid-template-columns: 1fr 1fr 1fr;
  }
}

@media screen and (min-width: 1458px) {
  .booksContainer {
    grid-template-columns: 1fr 1fr 1fr 1fr;
  }
}
