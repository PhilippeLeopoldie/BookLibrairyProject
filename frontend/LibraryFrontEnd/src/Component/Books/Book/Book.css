.bookcard--grid {
  border: 1px gray solid;
  display: grid;
  width: 18rem;
  height: 29rem;
  margin: 0 1rem 2rem 1rem;
  background-color: #ffffff;
  border-radius: 10px;
}

.bookcard--black {
  background-color: #000000;
  border: 1px #ffffff solid;
  .booktitle,
  .bookauthor {
    color: #ffffff;
  }
}


.boocard__Image {
  border-radius: 9px 9px 0 0;
  border-bottom: gray 1px solid;
  cursor: pointer;
}

.boocard__Image-zoomIn {
  transform: scale(1.2);
  border:gray 1px solid;
}

.boocard__Image-zoomOut {
  transform: scale(1);
}

.bookcard__footer--flex {
  display: flex;
  max-width: 18rem;
  align-items: center;
  justify-content: space-between;
  background-color: transparent;
}

.footer__average-rate--flex{
  display:flex;
  background-color: transparent;
}

.footer__average-rate {
  font-weight: bold;
  text-align: right;
  color: rgb(0, 76, 255);
  text-decoration: underline;
  cursor: pointer;
  background-color: transparent;
  margin-left: 10%;
}

.rate_star {
  margin-top: -1rem;
}

.bookCard__RateButton {
  margin-bottom: 4%;
  margin-right: 3%;
  width: 50%;
  font-size: 1rem;
  font-weight: bolder;
  color: rgb(0, 76, 255);
}

.bookcard__header {
  border-top-right-radius: 5%;
  border-top-left-radius: 5%;
  background-color: transparent;
}

.booktitle {
  background-color: transparent;
  font-size: 1.4rem;
  font-weight: bold;
  text-align: center;
  margin-top: 0;
  margin-bottom: 0.5rem;
  border-color: transparent;
  resize: none;
  max-height: 3.3rem;
  width: 94%;
  border-top-right-radius: 10%;
}

footer {
  background-color: transparent;
}

.bookauthor {
  background-color: transparent;
  margin-top: 0.2rem;
  margin-bottom: 0.2rem;
}

.bookform--view {
  resize: none;
}

.booktitle--trashbutton {
  display: flex;
  width: 25px;
  height: 30px;
  margin: 2%;
  padding-left: 0;
}

.bookcard--iconeTrash {
  border-radius: 20%;
  height: 25px;
}
