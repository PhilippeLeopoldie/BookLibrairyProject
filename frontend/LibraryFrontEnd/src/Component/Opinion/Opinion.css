.OpinionLoading {
  background-color: transparent;
}

.opinioncontainer {
  background-color: transparent;
  border-bottom-right-radius: 5%;
  border-bottom-left-radius: 5%;
  padding-left: 12%;
  padding-right: 3%;
  display: flex;
}

.opinioncontainer--card {
  display: grid;
  border-bottom-right-radius: 5%;
}

.opinionCardItems,
.img {
  background-color: rgb(253, 229, 187);
}

.opinioncard__reviews--flex {
  display: flex;
  justify-content: center;
}

.opinioncard__reviews__averageRate {
  font-weight: bold;
  text-align: right;
}

.opinioncard__reviews__star {
  text-align: left;
  margin-top: -1rem;
}

.OpinionCard__line {
  color: rgb(253, 229, 187);
  width: 90%;
  margin-bottom: 0;
}

.OpinionCard__LastReviewTitle {
  font-style: italic;
  text-align: left;
  font-size: 0.875rem;
  margin-bottom: 0.5rem;
}

.opinionCard__userName {
  text-align: left;
  font-weight: bold;
}

.opinioncard__rate--flex {
  display: flex;
}

.opinioncard--view {
  text-align: justify;
  font-style: italic;
  width: 17rem;
  height: 4rem;
  border-color: transparent;
  resize: none;
  overflow: hidden;
}

.opinioncard__postDate {
  margin-left: 0.5rem;
  text-align: left;
}

.opinioncard__nbReview {
  cursor: pointer;
  color: #4ba5ff;
  font-size: 1.125rem;
  font-weight: bold;
  text-align: left;
  margin: 0;
  text-decoration: underline;
}

.opinioncard--footer {
  display: flex;
  align-items: end;
  justify-content: center;
  border-bottom-right-radius: 5%;
}

.opinioncard--buttonmodify {
  margin-bottom: 5%;
}
.opinioncard--buttonmodify:disabled {
  opacity: 0.5;
  cursor: not-allowed;
}
