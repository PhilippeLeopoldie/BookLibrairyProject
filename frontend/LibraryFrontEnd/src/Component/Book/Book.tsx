import "./Book.css";
import { Opinion } from "../Opinion/Opinion";
import { OpinionCreate } from "../OpinionCreate/OpinionCreate";
import { OpinionList } from "../OpinionList/OpinionList";
import { ThemeContext } from "../App/App";
import trash from "../../media/delete.svg";
import url from "../../Url";
import { useContext, useState } from "react";

type BookWithDeletionHandler = {
  book?: {
    id: number;
    title: string;
    author: string;
    imageUrl?:string;
  };
  onDelete?: Function;
};

type Reviews = {
  id: number;
  rate: number;
  view: string;
  userName: string;
  postDate: string;
  bookId: number;
};

export const Book = ({ book, onDelete }: BookWithDeletionHandler) => {
  const theme = useContext(ThemeContext);
  const [createOpinionHandling, setCreateOpinionHandling] =
    useState<boolean>(false);
  const [displayReview, setDisplayReview] = useState<Boolean>(false);
  const [reviewList, setReviewList] = useState<Reviews[]>();

  const toggleOpinionList = () => {
    setDisplayReview(!displayReview);
  };
  const handleOpinionList = (reviews: Reviews[]) => {
    toggleOpinionList();
    setReviewList(reviews);
  };

  const toggleCreateOpinion = () => {
    setCreateOpinionHandling(!createOpinionHandling);
  };

  const DeleteBook = async (bookId: number) => {
    try {
      await fetch(url + `api/Book/${bookId}`, {
        method: "DELETE",
      });
      onDelete && onDelete(bookId);
    } catch (error) {
      console.error("Error deleting book:", error);
    }
  };

  return (
    <>
      {displayReview ? (
        <OpinionList
          opinions={reviewList}
          displayReviews={() => toggleOpinionList()}
          book={book}
        />
      ) : book && !createOpinionHandling ? (
        <div className={"bookcard--grid bookcard--" + theme}>
          <img
            src={book.imageUrl &&`${book.imageUrl}`}
            className="boocard__Image"
            alt="bookImage"
            width="288px"
            height="326px"
          />
          <header className="bookcard__header">
            {/* <button
            
              className=" button booktitle--trashbutton"
              type="submit"
              onClick={async () => {
                await DeleteBook(book.id);
              }}
            >
              <img
                className="icone bookcard--iconeTrash"
                src={trash}
                alt="delete"
              />
            </button> */}
            <h3  
              title="Book Title"
              className="booktitle">{book.title.length > 40 
                ? `${book.title.slice(0, 40)}...`
                : book.title}
            </h3>
            <h3 className="bookauthor">by: {book.author}</h3>
          </header>
          <footer className="bookcard__footer--flex">
          <Opinion
            book={book}
            toCreate={toggleCreateOpinion}
            displayReview={handleOpinionList}
          />
          
            <button
              className="button bookCard__RateButton"
              onClick={() => {
                toggleCreateOpinion();
              }}
            >
              Rate this book
            </button>
          </footer>
        </div>
      ) : (
        <OpinionCreate book={book && book} toCreate={toggleCreateOpinion} />
      )}
    </>
  );
};
