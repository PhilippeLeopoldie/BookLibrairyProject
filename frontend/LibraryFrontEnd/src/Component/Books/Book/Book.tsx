import "./Book.css";
import { Opinion } from "../../Opinions/Opinion/Opinion";
import { OpinionCreate } from "../../Opinions/OpinionCreate/OpinionCreate";
import { OpinionList } from "../../Opinions/OpinionList/OpinionList";
import { ThemeContext } from "../../App/App";
import { BOOK_BY_BOOKID_URL } from "../../../Url";
import { useContext, useEffect, useState } from "react";

type BookType = {
  book: {
    id: number;
    title: string;
    author: string;
    averageRate: number;
    imageUrl?: string;
  };
};

type Reviews = {
  id: number;
  rate: number;
  view: string;
  userName: string;
  postDate: string;
  bookId: number;
};

export const Book = ({ book }: BookType) => {
  const theme = useContext(ThemeContext);
  const [updatedBook, setUpdatedBook] = useState<BookType>({ book });
  const [displayCreateOpinion, setDisplayCreateOpinion] =
    useState<boolean>(false);
  const [opinionCreated, setOpinionCreated] = useState<boolean>(false);
  const [displayReview, setDisplayReview] = useState<Boolean>(false);
  const [averageClick, setAverageclick] = useState<Boolean>(false);
  const [reviewList, setReviewList] = useState<Reviews[]>();

  const fetchBook = async () => {
    try {
      const bookResponse: Response = await fetch(
        `${BOOK_BY_BOOKID_URL}${book?.id}`
      );
      if (bookResponse.status === 200) {
        const bookResponseData = await bookResponse.json();
        setUpdatedBook(bookResponseData);
      } else if (bookResponse.status === 404) {
        console.log(bookResponse);
      }
    } catch (error) {
      console.error(`Error fetching bookId:${book?.id}`, error);
    }
  };
  useEffect(() => {
    opinionCreated === true && fetchBook();
    console.log("opinionCreated status: ", opinionCreated);
  }, [opinionCreated]);

  const toggleOpinionList = () => {
    setDisplayReview(!displayReview);
    setAverageclick(false);
  };
  const handleOpinionList = (reviews: Reviews[]) => {
    toggleOpinionList();
    setReviewList(reviews);
  };

  const toggleCreateOpinion = () => {
    setDisplayCreateOpinion(!displayCreateOpinion);
  };

  const handleCreatedOpinion = (value: boolean) => {
    setOpinionCreated(value);
  };

  return (
    <>
      {displayReview ? (
        <OpinionList
          opinions={reviewList}
          displayReviews={() => toggleOpinionList()}
          book={book}
        />
      ) : book && !displayCreateOpinion ? (
        <div className={"bookcard--grid bookcard--" + theme}>
          <img
            src={book.imageUrl && `${book.imageUrl}`}
            className="boocard__Image"
            alt="bookImage"
            width="288px"
            height="326px"
          />
          <header className="bookcard__header">
            <h3 title="Book Title" className="booktitle">
              {book.title.length > 40
                ? `${book.title.slice(0, 40)}...`
                : book.title}
            </h3>
            <h3 className="bookauthor">by: {book.author}</h3>
          </header>
          <footer className="bookcard__footer--flex">
            <div className="footer__average-rate--flex">
              <a
                className="footer__average-rate"
                onClick={() => {
                  console.log("averageClick=", averageClick);
                  setAverageclick(!averageClick);
                }}
              >
                {book.averageRate}/5
              </a>
              <div className="rate_star"> &#9733;</div>
            </div>
            {averageClick && (
              <Opinion
                book={updatedBook?.book}
                displayReview={handleOpinionList}
              />
            )}
            
            <button
              className="button bookCard__RateButton"
              onClick={() => {
                toggleCreateOpinion();
              }}
            >
              Rate this book
            </button>
          </footer>
          
        </div>
      ) : (
        <OpinionCreate
          book={book && book}
          toCreate={toggleCreateOpinion}
          created={handleCreatedOpinion}
        />
      )}
    </>
  );
};
