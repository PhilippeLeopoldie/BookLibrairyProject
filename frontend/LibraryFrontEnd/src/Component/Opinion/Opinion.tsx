import { useEffect, useState, useContext } from "react";
import "./Opinion.css";
import { OpinionList } from "../OpinionList/OpinionList";
import url from "../../Url";
import modify from "../../media/write.svg";
import { Rate } from "../Rate/Rate";

type OpinionType = {
  id: number;
  rate: number;
  view: string;
  userName: string;
  postDate: string;
  bookId: number;
};

type BookType = {
  book: {
    id: number;
    title: string;
    author: string;
  };
  onEdit?: (opinionToUpdate: OpinionType) => void;
  toCreate: () => void;
  displayReview: (opinions: OpinionType[]) => void;
};

export const Opinion = ({
  book,
  onEdit,
  toCreate,
  displayReview,
}: BookType) => {
  const [opinions, setOpinions] = useState<OpinionType[] | null>(null);
  const [fetching, setFething] = useState<boolean>(true);
  const [lastOpinion, setLastOpinion] = useState<OpinionType | null>(null);
  const [averageRate, setAverageRate] = useState<number>(0);
  const FetchOpinions = async (bookId: number) => {
    try {
      const response: Response = await fetch(
        url + "api/Opinion/BookId=" + bookId
      );
      if (response.status === 200) {
        const responseData = await response.json();
        setOpinions(responseData.$values);
        setFething(false);
      } else if (response.status === 404) {
        setFething(false);
        console.log("404 response", response);
      }
    } catch (error) {
      console.error("error fetching Opinions:", error);
    }
  };

  const AverageRate = () => {
    if (opinions) {
      const rateSum = opinions.reduce(
        (total, opinion) => total + opinion.rate,
        0
      );
      const average = rateSum / opinions.length;

      const formattedAverage = Number(average.toFixed(1));
      setAverageRate(formattedAverage);
    }
  };

  useEffect(() => {
    FetchOpinions(book.id);
  }, []);

  useEffect(() => {
    opinions &&
      opinions.length > 0 &&
      setLastOpinion(opinions[opinions.length - 1]);
    AverageRate();
    console.log("average rate: ", averageRate);
  }, [opinions]);

  if (fetching) {
    return <h2 className="OpinionLoading">Loading...</h2>;
  }

  return (
    <>
      <div className="opinioncontainer">
        {lastOpinion && (
          <div className="opinioncontainer--card" key={lastOpinion.id}>
            <div className="opinionCardItems">
              <div className="opinionCardItems opinioncard__reviews--flex">
                <a
                  className="opinionCardItems opinioncard__nbReview"
                  onClick={() => {
                    opinions && displayReview(opinions);
                  }}
                >
                  {opinions &&
                    (opinions.length > 1
                      ? `${opinions.length} reviews`
                      : `${opinions.length} review`)}
                </a>
                <div className="opinionCardItems opinioncard__reviews__averageRate">
                  {averageRate}
                </div>
                <div className="rate_star opinioncard__reviews__star">
                  &#9733;
                </div>
              </div>
              <hr className="OpinionCard__line"></hr>
              <div className="opinionCardItems opinionCard__userName">
                {lastOpinion.userName}
              </div>
              <div className="opinionCardItems opinioncard__rate--flex">
                <Rate rate={lastOpinion.rate} />
                <p className="opinionCardItems opinioncard__postDate">
                  {lastOpinion.postDate}
                </p>
              </div>
              <textarea
                className="opinionCardItems opinioncard--view"
                value={
                  lastOpinion.view.length > 150
                    ? `${lastOpinion.view.slice(0, 110)}...`
                    : lastOpinion.view
                }
                readOnly
              />
            </div>
          </div>
        )}
      </div>
    </>
  );
};
