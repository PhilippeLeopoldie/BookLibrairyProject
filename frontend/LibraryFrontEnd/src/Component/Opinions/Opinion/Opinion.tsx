import { useEffect, useState } from "react";
import "./Opinion.css";
import { OPINION_BY_BOOKID_URL } from "../../../constants/api";


type OpinionType = {
    id: number;
    rate: number;
    view: string;
    userName: string;
    postDate: string;
    bookId: number;
};

type BookType = {
    book: {
        id: number;
        title: string;
        author: string;
        imageUrl?: string;
        averageRate?: number;
    };
    displayReview: (opinions: OpinionType[]) => void;
};

export const Opinion = ({book,displayReview}: BookType) => {
    
    const FetchOpinionBYBookId = async (bookId: number) => {
        try {
            const response: Response = await fetch(
                OPINION_BY_BOOKID_URL + bookId
            );
            if (response.status === 200) {
                const responseData = await response.json();
                displayReview(responseData);
            } else if (response.status === 404) {
                console.log("404 response", response);
            }
        } catch (error) {
            console.error("error fetching Opinions:", error);
        }
    };

    useEffect(() => {
        FetchOpinionBYBookId(book.id);
    }, []);

    return (
        <>
        </>
    );
};
