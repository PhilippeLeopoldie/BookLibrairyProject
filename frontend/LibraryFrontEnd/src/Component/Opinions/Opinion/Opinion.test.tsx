import { Opinion } from "./Opinion";
import { act, render, screen } from "@testing-library/react";
import "@testing-library/jest-dom";
import { Book } from "../../Books/Book/Book";

global.fetch = jest.fn().mockImplementation(() =>
  Promise.resolve({
    json: () =>
      Promise.resolve([
        { Opinion: { id: 1, view: "View 1", userName: "userName 1", rate: 2 } },
      ]),
    status: 200,
    hearders: new Headers(),
  })
);

type Book={
  id:number,
  title: string,
  author:string
}
const book :Book = {
  id:1,
  title:"title 1",
  author: "author 1"
}

describe("Opinions Component Tests", () => {
  it("SnapShot: should render Opinion", () => {
    const {asFragment} =  
      render(<Opinion 
        book={book} 
        onEdit={() => {}} 
        //toCreate={()=> {}} 
        displayReview={()=>{}}/>);
    expect(asFragment()).toMatchSnapshot();
  })

  it("Should not render Loading message when book with no opinion", async () => {
    //Arrange
    
    await act(async () => {
      render(<Opinion 
        book={book} 
        onEdit={() => {}} 
        //toCreate={()=> {}}
        displayReview={()=>{}} />);
    });
  
    //Act
    const loadingMessage = screen.queryByText("Loading...");
  
    //Expect
    expect(loadingMessage).toBeNull();
  }); 
})

