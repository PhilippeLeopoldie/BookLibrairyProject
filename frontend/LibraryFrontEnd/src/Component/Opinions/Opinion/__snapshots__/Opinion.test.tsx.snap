// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Opinions Component Tests SnapShot: should render Opinion 1`] = `
<DocumentFragment>
  <h2
    class="OpinionLoading"
  >
    Loading...
  </h2>
</DocumentFragment>
`;
